torch/distributed/device_mesh.py

```diff
@@ -1021,20 +801,36 @@ def _flatten(
             After the flattened dimension is created, to access the flattened dimension in mesh_3d, one can use the
             existing slicing method to obtain the flattened mesh through calling mesh_3d["dp_cp"].
             """
-            if not self.mesh_dim_names:
-                raise RuntimeError(
-                    "Cannot flatten a DeviceMesh without mesh_dim_names!"
-                )
-
             if backend_override is not None:
                 (backend_override_tuple,) = _normalize_backend_override(
                     {0: backend_override}, 1
                 )
             else:
                 backend_override_tuple = (None, None)
 
-            return _mesh_resources.create_flatten_mesh(
-                self, mesh_dim_name, backend_override_tuple
+            global_override = _mesh_resources.mesh_dim_group_options.get(0, (None, None))
+            if backend_override_tuple == (None, None):
+                backend_override_tuple = global_override
+            elif backend_override_tuple != (None, None):
+                raise RuntimeError(
+                    "Dimension 0 present both in the backend_override argument "
+                    "and via _mesh_resources._set_mesh_dim_group_options"
+                )
+
+            flattened_layout = coalesce(
+                Layout(
+                    sorted(
+                        [ss for l in self.layouts for ss in l.sizes_and_strides], key=lambda x: x[1]
+                    )
+                )
+            )
+
+            self.storage._maybe_create_dim(
+                flattened_layout, mesh_dim_name, backend_override=backend_override_tuple
```

Comment:
you need to borrow logic from `create_flatten_mesh` to get fallback dim_name since this is optional...